src/Hosting/SystemdServices/src/ServiceHelpers.cs

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.Extensions.Hosting.SystemdServices
+{
+    /// <summary>
+    /// Helper methods for systemd Services.
+    /// </summary>
+    public static class ServiceHelpers
+    {
+        const string INVOCATION_ID = "INVOCATION_ID";
+
+        private static string s_invocationId;
+
+        /// <summary>
+        /// Check if the current process is hosted as a systemd Service.
+        /// </summary>
+        /// <returns><c>True</c> if the current process is hosted as a systemd Service, otherwise <c>false</c>.</returns>
+        public static bool IsSystemdService()
+        {
+            // We use the INVOCATION_ID envirionment variable that was introduced in systemd 232 (released 2016-11-03).
```

Comment:
This seems so wrong to me. Why do we need to care about child processes not having this environment variable?

Reply:
If an env-var is the only way we can detect this, we might be stuck with this option.

For windows services, [we check the identity of the parent process](https://github.com/aspnet/Extensions/blob/9d6a021d72c2495d0944c4557c1847181d545061/src/Hosting/WindowsServices/src/ServiceHelpers.cs#L25). Is that a viable option here? IIRC, it's not trivial to detect your parent process in Linux though... Looks like it can be [read from `/proc/self/stat`](http://man7.org/linux/man-pages/man5/proc.5.html) if there isn't already an API in .NET Core for it.

Reply:
> This seems so wrong to me. Why do we need to care about child processes not having this environment variable?

The corresponding WindowsService function also won't return true for children.

> r windows services, we check the identity of the parent process. Is that a viable option here? IIRC, it's not trivial to detect your parent process in Linux though... Looks like it can be read from /proc/self/stat if there isn't already an API in .NET Core for it.

There are a number of options, the envvar is the simplest.
Other option: look at the parents, use cgroup names.
note: we want this to work also for systemd instances that run per user, that adds some complexity.